### PR TITLE
Make dart_lock thread-safe

### LINKED ISSUE #320: DART locality subsystem initialized multiple times when thread support is disabled
See logs attached.

[test.nothreads.txt](https://github.com/dash-project/dash/files/838088/test.nothreads.txt)
[test.threads.txt](https://github.com/dash-project/dash/files/838087/test.threads.txt)

### LINKED ISSUE #330: Fix thread-safety of dart_lock
After having a closer look at the implementation of `dart_lock` in `dart_synchronization.c`, I think this needs fixing:

1. The implementation of `dart_lock_acquire` and `dart_lock_release` is not thread-safe. This can be fixed through a local lock quite easily. Creation and destruction will not be made thread-safe though (as it is a collective operation). 
2. ~~More important though: there is a race condition and potential deadlock in dart_lock_acquire. If the unit holding the lock releases it after a second unit determined that the lock is not free (first fetch_and_op) and before the second unit registered in the first unit's tail pointer (second fetch_and_op). We may need to resort to MPI_Win_lock, thus circumventing the allocation mechanisms in DART (which is fine since it is still dart-mpi).~~
UPDATE: the code is correct in this respect. I just had to massage it a bit and make it better readable before I could fully understand it :)

I will try to come up with an improved implementation soon.

### PR DESCRIPTION
**Pull request by devreal**
Major cleanups in `dart_synchronization.c` and added thread-safety. Also a few minor fixes.

Fixes #330.

Also fixes:
1. double-execution of the `expr` passed to `DART_ASSERT_RETURNS`.
2. Globally setting release flags if thread support has been detected.  
Fixes #320.

### CODE REVIEW

File: dart-impl/base/include/dash/dart/base/mutex.h [hunk 1]
```diff
@@ -45,7 +43,28 @@ dart__base__mutex_init(dart_mutex_t *mutex)
...
+  if (ret != 0) {
+    DART_LOG_WARN("dart__base__mutex_init_recursive: Failed to initialize "
+                  "recursive pthread mutex! Falling back to fast mutex...");
+    pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_FAST_NP);
+  }
+  pthread_mutex_init(&mutex->mutex, &attr);
+  pthread_mutexattr_destroy(&attr);
+  return DART_OK;
+#else
+  return DART_ERR_INVAL;
```
Comment (fuchsto) [inline dart-impl/base/include/dash/dart/base/mutex.h]: Add a log error message here

File: dart-impl/mpi/src/dart_synchronization.c [hunk 2]
```diff
@@ -25,252 +27,328 @@
...
+    DART_TYPE_BYTE,
+    DART_TEAM_UNIT_ID(0),
+    teamid);
+
+  /* Create a global memory region across the team.
+   * Every local memory segment holds the next unit
+   * waiting on the lock. */
+  if (DART_OK !=
+        dart_team_memalloc_aligned(teamid, 1, DART_TYPE_INT, &gptr_list)) {
+    return DART_ERR_OTHER;
```
Comment (fuchsto) [inline dart-impl/mpi/src/dart_synchronization.c]: Should return error code from `dart_team_memalloc_aligned`

File: dart-impl/mpi/src/dart_synchronization.c [hunk 3]
```diff
@@ -25,252 +27,328 @@
...
-dart_ret_t dart_lock_acquire (dart_lock_t lock)
+dart_ret_t dart_lock_acquire(dart_lock_t lock)
 {
+  /* lock the local mutex and keep it until the global lock is released */
+  dart__base__mutex_lock(&lock->mutex);
+
+  if (lock->is_acquired == 1)
+  {
+    DART_LOG_ERROR("dart_lock_acquire: LOCK has already been acquired\n");
+    dart__base__mutex_unlock(&lock->mutex);
```
Comment (fuchsto) [inline dart-impl/mpi/src/dart_synchronization.c]: All return values unchecked

File: dart-impl/mpi/src/dart_synchronization.c [hunk 4]
```diff
@@ -25,252 +27,328 @@
...
-  MPI_Aint disp_list;
+  uint64_t    tail_offset = gptr_tail.addr_or_offs.offset;
+  dart_unit_t tail_unit   = gptr_tail.unitid;
 
   dart_team_data_t *team_data = dart_adapt_teamlist_get(lock->teamid);
   if (team_data == NULL) {
-    DART_LOG_ERROR("dart_lock_acquire ! failed: Unknown team %i!", lock->teamid);
+    DART_LOG_ERROR("dart_lock_acquire ! failed: Unknown team %i!",
+                   lock->teamid);
+    dart__base__mutex_unlock(&lock->mutex);
```
Comment (fuchsto) [inline dart-impl/mpi/src/dart_synchronization.c]: Return value unchecked

File: dart-impl/mpi/src/dart_synchronization.c [hunk 5]
```diff
@@ -25,252 +27,328 @@
...
+    *is_acquired = 0;
+    return DART_OK;
+  }
+
+  if (lock->is_acquired == 1)
+  {
+    DART_LOG_ERROR("dart_lock_try_acquire: LOCK has already been acquired\n");
+    *is_acquired = 1;
+    /* we are using a recursive lock so give up this recursion */
+    dart__base__mutex_unlock(&lock->mutex);
```
Comment (fuchsto) [inline dart-impl/mpi/src/dart_synchronization.c]: Return value unchecked

File: dart-impl/mpi/src/dart_synchronization.c [hunk 6]
```diff
@@ -25,252 +27,328 @@
...
+  /* If the old predecessor was -1, we have claimed the lock,
+   * otherwise, do nothing. */
+  if (result == -1)
+  {
+    lock->is_acquired = 1;
+    *is_acquired = 1;
+  } else {
+    *is_acquired = 0;
+    /* unlock the local mutex if we have not acqcuired the global lock */
+    dart__base__mutex_unlock(&lock->mutex);
```
Comment (fuchsto) [inline dart-impl/mpi/src/dart_synchronization.c]: Return value unchecked

File: dart-impl/mpi/src/dart_synchronization.c [hunk 7]
```diff
@@ -25,252 +27,328 @@
...
-
     MPI_Send(NULL, 0, MPI_INT, next, 0, team_data->comm);
-    *addr2 = -1;
+    *addr = -1;
     MPI_Win_sync(win);
   }
-  lock -> is_acquired = 0;
-  DART_LOG_DEBUG("UNLOCK - release lock in team %d", (lock -> teamid));
+  lock->is_acquired = 0;
+  dart__base__mutex_unlock(&lock->mutex);
```
Comment (fuchsto) [inline dart-impl/mpi/src/dart_synchronization.c]: Return value unchecked

File: dart-impl/mpi/src/dart_synchronization.c [hunk 8]
```diff
@@ -25,252 +27,328 @@
...
+  dart_team_myid(teamid, &unitid);
+  if (unitid.id == 0)
+  {
+    dart_memfree(gptr_tail);
+  }
+
+  dart_team_memfree(gptr_list);
+  (*lock)->gptr_tail = DART_GPTR_NULL;
+  (*lock)->gptr_list = DART_GPTR_NULL;
+  dart__base__mutex_destroy(&(*lock)->mutex);
```
Comment (fuchsto) [inline dart-impl/mpi/src/dart_synchronization.c]: All return values unchecked in this function

**Review by fuchsto — CHANGES_REQUESTED** [review]

**Comment by devreal** [comment]
@fuchsto Should all be fixed.

File: dart-impl/base/include/dash/dart/base/mutex.h [hunk 9]
```diff
@@ -33,64 +31,105 @@ char __dummy;
...
 
-static inline
+DART_INLINE
 dart_ret_t
 dart__base__mutex_init(dart_mutex_t *mutex)
 {
 #ifdef DART_HAVE_PTHREADS
-  pthread_mutex_init(&mutex->mutex, NULL);
+  // pthread_mutex_init always succeeds
+  pthread_mutex_init(&mutex->mutex,
```
Comment (fmoessbauer) [inline dart-impl/base/include/dash/dart/base/mutex.h]: Does this compile? Syntax looks faulty to me.

Comment (devreal) [reply dart-impl/base/include/dash/dart/base/mutex.h]: Oh boy, my touchpad tricked me into that. Thanks for spotting this, fixed a bunch of other problems as well.

**Review by fuchsto — APPROVED** [review]

**Comment by devreal** [comment]
The thread-safety tests fail for strange reasons. I am on it but so far clueless...

**Comment by fmoessbauer** [comment]
I guess the pointer is not initialized correctly?

```
[    2 ERROR ] [  1959 ] dart_globmem.c           :162  !!! DART: dart_memfree: invalid local global pointer: invalid offset: 0
[    2 ERROR ] [  1959 ] dart_synchronization.c   :374  !!! DART: Failed to free global mmeory
terminate called after throwing an instance of 'dash::exception::AssertionFailed'
  what():  [ Unit 2 ] Assertion 0 == 2 failed: dart_team_lock_free failed /opt/dash/dash/src/Mutex.cc:14
[    1 ERROR ] [  1958 ] dart_globmem.c           :162  !!! DART: dart_memfree: invalid local global pointer: invalid offset: 0
[    1 ERROR ] [  1958 ] dart_synchronization.c   :374  !!! DART: Failed to free global mmeory
```

File: dart-impl/mpi/src/dart_synchronization.c [hunk 10]
```diff
@@ -25,252 +27,365 @@
...
+    DART_ASSERT_RETURNS(
+      MPI_Win_flush(predecessor, win),
+      DART_OK);
+
+    /* Waiting for notification from its predecessor*/
+    DART_LOG_DEBUG("dart_lock_acquire: waiting for notification from "
+                   "%d in team %d",
+                   predecessor, lock->teamid);
+
+    MPI_Recv(NULL, 0, MPI_INT, predecessor, 0, team_data->comm,
```
Comment (fmoessbauer) [inline dart-impl/mpi/src/dart_synchronization.c:221]: This seems to hang, at least in debug builds. Unit 0 waits for itself.
```
[    1 DEBUG ] [ 32334 ] Shared.h                 :86   | Shared.Shared(team,owner) >                  | |- _dart_gptr: dart_gptr_t(<000000|00|0000|0000|0000000000000008>)
[    0 DEBUG ] [ 32333 ] dart_communication.c     :781   DART: dart_put_blocking > finished
[    2 TRACE ] [ 32335 ] dart_segment.c           :168   DART: dart_segment_get_disp() seq_id:1 rel_unitid:2
[    2 TRACE ] [ 32335 ] dart_segment.c           :181   DART: dart_segment_get_disp > disp:94157135832688
[    2 TRACE ] [ 32335 ] dart_segment.c           :168   DART: dart_segment_get_disp() seq_id:1 rel_unitid:2
[    2 TRACE ] [ 32335 ] dart_segment.c           :181   DART: dart_segment_get_disp > disp:94157135832688
[    2 DEBUG ] [ 32335 ] dart_synchronization.c   :206   DART: dart_lock_acquire: waiting for notification from 2 in team 0
[    1 TRACE ] [ 32334 ] dart_segment.c           :168   DART: dart_segment_get_disp() seq_id:1 rel_unitid:1
[    1 TRACE ] [ 32334 ] dart_segment.c           :181   DART: dart_segment_get_disp > disp:94914551350896
[    1 TRACE ] [ 32334 ] dart_segment.c           :168   DART: dart_segment_get_disp() seq_id:1 rel_unitid:1
[    1 TRACE ] [ 32334 ] dart_segment.c           :181   DART: dart_segment_get_disp > disp:94914551350896
[    1 DEBUG ] [ 32334 ] dart_synchronization.c   :206   DART: dart_lock_acquire: waiting for notification from 1 in team 0
[    0 TRACE ] [ 32333 ] dart_segment.c           :168   DART: dart_segment_get_disp() seq_id:1 rel_unitid:0
[    0 TRACE ] [ 32333 ] dart_segment.c           :181   DART: dart_segment_get_disp > disp:94115753625456
[    0 TRACE ] [ 32333 ] dart_segment.c           :168   DART: dart_segment_get_disp() seq_id:1 rel_unitid:0
[    0 TRACE ] [ 32333 ] dart_segment.c           :181   DART: dart_segment_get_disp > disp:94115753625456
[    0 DEBUG ] [ 32333 ] dart_synchronization.c   :206   DART: dart_lock_acquire: waiting for notification from 0 in team 0
```

Comment (devreal) [reply dart-impl/mpi/src/dart_synchronization.c:221]: This is the same that @rkowalewski reported in #339. However, I cannot reproduce this locally. What is your test case? One node? Multi-node? 

I added two test cases to specifically test the `dart_lock` implementation: https://github.com/dash-project/dash/blob/bug-330-dart-lock/dash/test/dart/DARTLockTest.cc

Can you please run this basic test to verify correctness of DART log?

**Comment by devreal** [comment]
I have fixed an invalid free in `dart_team_lock_destroy`, yes.

File: dart-impl/mpi/src/dart_synchronization.c (hunk 10 above)
Comment (fmoessbauer) [reply dart-impl/mpi/src/dart_synchronization.c:221]: Hi, this output was generated using a debug build, 1 node, 3 units, OpenMPI 1.10.3:
[dart_mutex.txt](https://github.com/dash-project/dash/files/875093/dart_mutex.txt)

Comment (devreal) [reply dart-impl/mpi/src/dart_synchronization.c:221]: OK, so this was a tough one: 

1. The reason this started to fail: I added error checking to the `dart_lock` interface. Sensible thing. 
2. The root cause: `DART_ASSERT_RETURNS` called the `expr` twice, once in the `if` and once in `assert`. Of course, calling `MPI_Fetch_and_op` twice does not yield the same  result on the second call as on the first call. 

I think someone here owes me a beer :)

**Comment by devreal** [comment]
This might also solve #320. I couldn't reproduce this issue with thread-support enabled, for reasons I have not figured out, yet.

**Comment by devreal** [comment]
Yep, this should do it: if thread-support is detected, the code in `Threadsupport.cmake` falsely overwrote `CMAKE_C_FLAGS` with `CMAKE_C_FLAGS_RELEASE`, which set `-DNDEBUG` globally and thus disabled the `assert` in the `DART_ASSERT*` macros. Should be all fixed now.

**Comment by fmoessbauer** [comment]
Great effort!!! That killed several birds with one stone hit.

**Comment by fuchsto** [comment]
Good stuff as usual, but also close to no unit tests as usual :P
Could you extend them? Hard to believe the additional 100 lines of test cover the changes in this PR.

**Review by fuchsto — CHANGES_REQUESTED** [review]
@devreal Good stuff as usual, but also close to no unit tests as usual :P
Could you extend them? Hard to believe the additional 100 lines of test cover the changes in this PR.

EDIT: Naah, I owe you one. Will add more tests and merge.

**Comment by devreal** [comment]
I improved test coverage and adapted the interface a bit, including documentation. Let me know if you can think of any other test that is missing.

**Review by fuchsto — APPROVED** [review]

**Comment by fuchsto** [comment]
Thank you, that was quick.

**Pull request merged**